### PR TITLE
Pix2Struct onnxruntime support

### PR DESCRIPTION
**Pull request by krathul**
This PR implements onnxruntime support for Pix2struct using ORT class (#1288)

### CODE REVIEW

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 1]
```diff
@@ -1362,3 +1455,217 @@ def _reorder_cache(past, beam_idx) -> Tuple[Tuple[torch.FloatTensor]]:
...
+        # Encode if needed : first prediction pass
+        # Encode if needed (training, first prediction pass)
+        if encoder_outputs is None:
+            encoder_outputs = self.encoder(
+                flattened_patches=flattened_patches,
+                attention_mask=attention_mask,
+            )
+        # work around for Unexpected input data type (tensor(float)) for attention_mask
+        # Need to be looked into
+        attention_mask = attention_mask.to(torch.int64)
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:1520]: Did you find the issue?

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 2]
```diff
@@ -1362,3 +1455,217 @@ def _reorder_cache(past, beam_idx) -> Tuple[Tuple[torch.FloatTensor]]:
...
+    """
+    Pix2struct model with a language modeling head for ONNX Runtime inference.
+    """
+
+    # pix2struct cannot be loaded using AutoModel
+    auto_model_class = Pix2StructForConditionalGeneration
+    main_input_name = "flattened_patches"
+
+    # Copied from transformers.models.t5.modeling_t5.T5PreTrainedModel._shift_right with T5->Pix2Struct
+    def _shift_right(self, input_ids):
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:1470]: @JingyaHuang In transformers, this seem to be used for training. I never really looked at the ONNX config with loss, but does this include a `label` input? i.e., in ORTTrainer, is the export going through e.g. https://github.com/huggingface/transformers/blob/e42587f596181396e1c4b63660abf0c736b10dae/src/transformers/models/t5/modeling_t5.py#L1701-L1703 ?

@krathul I think for now it is reasonable to remove this method. We can have a look at training later.

**Review by fxmarty — COMMENTED** [review]
Thank you, it looks in good shape! I'll add tests in this PR if that is fine to you.

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 2 above)
Comment (krathul) [reply optimum/onnxruntime/modeling_seq2seq.py:1469]: Ok

**Comment by krathul** [comment]
@fxmarty you can add tests.

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 1 above)
Comment (krathul) [reply optimum/onnxruntime/modeling_seq2seq.py]: Yes, the attention_mask returned by pix2structProcessor is of type torch.float32, it is converted for use in `ORTEncoderForPix2Struct` but is directly used in `ORTDecoderForSeq2Seq` for the decoder, which requires a long tensor. As a solution, I guess the conversion can be done in forward method since the decoder method is generic.

Comment (fxmarty) [reply optimum/onnxruntime/modeling_seq2seq.py]: Yes, we should fix the dtype of the attention mask in the ONNX export. As a dirty workaround I kept yours, but it is not ideal.

File: optimum/exporters/onnx/model_configs.py [hunk 3]
```diff
@@ -1275,12 +1284,46 @@ def inputs(self):
...
         return common_inputs
 
+    @property
+    def outputs(self) -> Dict[str, Dict[int, str]]:
+        if self._behavior is ConfigBehavior.ENCODER:
+            common_outputs = {
+                "last_hidden_state": {0: "batch_size"}
+            }  # The last hidden state dim=1 is constant, no need for it to be dynamic.
+        else:
+            common_outputs = super(OnnxConfigWithPast, self).outputs
```
Comment (fxmarty) [inline optimum/exporters/onnx/model_configs.py:1301]: We redefine because of this comment.

**Comment by fxmarty** [comment]
The PR is in good shape. On my CPU I am getting a 17% speedup over pytorch.

**Review by fxmarty — APPROVED** [review]
@krathul Thanks you for the clean implementation!

**Comment by krathul** [comment]
> @krathul Thank you for the clean implementation!

@fxmarty Thanks for helping with the merge
